Safer Shell command Execution

mattKorwel opened the pull request:
## TLDR

  This pull request enhances the run_shell_command tool by improving command parsing and adding better security checks. Key changes include stripping shell wrappers (like sh -c and
  cmd.exe /c), blocking command and process substitution, and providing more detailed confirmation prompts for multi-part commands.

## Dive Deeper

  The primary goal of this PR is to make the run_shell_command tool safer and more robust. The existing implementation had several limitations that could be bypassed with more complex
  shell commands.

  The main changes are:

   - `getCommandRoots`: This function is now more comprehensive and can correctly identify the root commands in a chained command, including those using &.
   - `stripShellWrapper`: This new function removes common shell wrappers from the command string, allowing for more accurate parsing and validation. It handles sh, bash, zsh, and cmd.exe.
   - `isCommandAllowed`: This function now blocks process substitution (<() and >()) in addition to command substitution ($(...)).
   - `ToolConfirmationMessage.tsx`: The confirmation prompt for shell commands now lists all root commands that will be executed, giving the user a clearer understanding of what they are
     approving.

  These changes have been tested on both Unix-like shells and the Windows Command Prompt.

  Reviewer Test Plan

  To test these changes, you can try running shell commands with various levels of complexity. Here are some examples:

   - Simple command: ls -l
   - Chained command: echo "hello" && echo "world"
   - Piped command: ls -l | grep "test"
   - Command with `sh -c`: sh -c "ls -l"
   - Command with `cmd.exe /c` (on Windows): cmd.exe /c "dir"
   - Blocked command (command substitution): echo $(ls)
   - Blocked command (process substitution): diff <(ls) <(ls -a)

  When running these commands, you should observe the following:

   - The confirmation prompt should correctly identify the root commands.
   - Allowed commands should execute as expected.
   - Blocked commands should be rejected with a security warning.


## UX Before and After

<img width="1388" height="310" alt="Screenshot 2025-07-24 at 11 10 51 AM" src="https://github.com/user-attachments/assets/bbf18dbc-50e4-4ab7-bc45-43682b6bf775" />
<img width="1353" height="250" alt="Screenshot 2025-07-24 at 11 11 53 AM" src="https://github.com/user-attachments/assets/6fceebad-82ed-4cac-9979-1ca0abc432eb" />
<img width="1383" height="317" alt="Screenshot 2025-07-24 at 11 13 44 AM" src="https://github.com/user-attachments/assets/a3835e64-0a5d-4093-b022-e05a4523d33a" />
<img width="1375" height="304" alt="Screenshot 2025-07-24 at 11 14 15 AM" src="https://github.com/user-attachments/assets/88c057c2-e804-47be-882a-593ba1a27469" />
<img width="2780" height="1220" alt="image" src="https://github.com/user-attachments/assets/992951e2-7aab-46d3-9678-a1d6c16a7000" />


## Testing Matrix


  ┌──────────┬────┬────┬────┐
  │          │ 🍏 │ 🪟 │ 🐧 │
  ├──────────┼────┼────┼────┤
  │ npm run  │ ✅ │ ✅ │ ✅ │
  │ npx      │ ❓ │ ❓ │ ❓ │
  │ Docker   │ ❓ │ ❓ │ ❓ │
  │ Podman   │ ❓ │ -  │ -  │
  │ Seatbelt │ ❓ │ -  │ -  │
  └──────────┴────┴────┴────┘


## Linked issues / bugs

  This PR does not address any specific issues.


## Disclosure

@sam-cox-tracebit and @thebigsasha (https://github.com/google-gemini/gemini-cli/issues/2106)

NTaylorMullen commented on packages/core/src/tools/shell.ts:
```diff
@@ -120,11 +146,15 @@ Process Group PGID: Process group started or \`(none)\``,
+    if (
+      command.includes('$(') ||
+      command.includes('<(') ||
+      command.includes('>')
```
Did you mean for this to have a paren after it?

mattKorwel commented on packages/cli/src/ui/components/messages/ToolConfirmationMessage.tsx:
```diff
@@ -132,18 +132,25 @@ export const ToolConfirmationMessage: React.FC<
-        label: `Yes, allow always "${executionProps.rootCommand} ..."`,
+    const isMultiCommand = executionProps.rootCommand.includes(',');
+    question = `Allow execution of ${
+      isMultiCommand ? 'these commands' : 'this command'
```
added.

mattKorwel commented on packages/cli/src/ui/components/messages/ToolConfirmationMessage.tsx:
```diff
@@ -132,18 +132,25 @@ export const ToolConfirmationMessage: React.FC<
-      {
-        label: `Yes, allow always "${executionProps.rootCommand} ..."`,
+    const isMultiCommand = executionProps.rootCommand.includes(',');
+    question = `Allow single execution of ${
```
```suggestion
    question = `Allow execution ${
```

smhendrickson commented on packages/core/src/tools/shell.ts:
```diff
@@ -259,18 +289,25 @@ Process Group PGID: Process group started or \`(none)\``,
+      rootCommand: commandsToConfirm.join(', '),
       onConfirm: async (outcome: ToolConfirmationOutcome) => {
         if (outcome === ToolConfirmationOutcome.ProceedAlways) {
-          this.whitelist.add(rootCommand);
```
Oh i see we are using the presence of ','. This works.

mattKorwel commented:
This pull request enhances the shell tool by improving command parsing and security. It now correctly identifies and handles chained commands (e.g., `git status && git log`), disallows certain types of command substitution for security, and provides clearer user confirmation prompts. The changes also include significant refactoring and expansion of the test suite to ensure robustness.

mattKorwel commented:
Summary

  I've successfully updated the command parsing logic to follow proper bash quoting rules as you specified:

  Updated Command Substitution Detection

  The detectCommandSubstitution method now correctly handles:

  Single quotes ('): Everything literal - no command substitution possible
  - gh pr create --body 'This is \code and $(not a command)'` → SAFE

  Double quotes ("): Command substitution with $() and backticks unless escaped
  - gh pr create --body "This has $(dangerous) content" → BLOCKED
  - gh pr create --body "This has \dangerous content"` → BLOCKED
  - gh pr create --body "This has \\$(safe) content" → SAFE (escaped)

  No quotes: Command substitution with $(), <(), and backticks
  - echo $(dangerous) → BLOCKED
  - echo \dangerous` → BLOCKED
  - diff <(ls) <(ls -a) → BLOCKED

  All Tests Passing (64/64)

  Core Security Tests:
  - Blocks unquoted $(...) command substitution
  - Blocks unquoted <(...) process substitution
  - Blocks unquoted backticks (corrected from previous insecure behavior)
  - Allows patterns within single quotes (safe)
  - Blocks patterns within double quotes (dangerous)
  - Allows properly escaped patterns
  - Handles complex multi-line content safely

  Best Practices for Markdown

  For commands like gh that accept markdown content, the safest approaches are:

  # RECOMMENDED: Single quotes (safest)
  gh pr create --body 'This is `code` and $(not executed)'

  # ALTERNATIVE: Escaped double quotes
  gh pr create --body "This is \`code\` and \$(not executed)"

  # BEST: Here documents for complex content
  gh pr create --body "$(cat <<'EOF'
  # My PR
  This has `code blocks` and $(variables) that won't be interpreted
  EOF
  )"

  The implementation now correctly follows bash security principles while maintaining usability for legitimate markdown use cases.

mattKorwel commented:
For git commit messages, backticks work the same way as other bash contexts:

  Direct command line:
  # DANGEROUS - executes `ls` command
  git commit -m "Fix `ls` output formatting"

  # SAFE - backticks treated literally
  git commit -m 'Fix `ls` output formatting'

  # SAFE - escaped in double quotes
  git commit -m "Fix \`ls\` output formatting"

  Using heredoc (recommended for multi-line):
  git commit -m "$(cat <<'EOF'
  Fix code formatting

  - Updated `formatOutput()` function
  - Fixed `processData()` method
  - Added tests for `validateInput()`
  EOF
  )"

  Git's -F flag with temp file:
  # Write to temp file first
  cat > /tmp/commit_msg <<'EOF'
  Fix code formatting

  - Updated `formatOutput()` function
  - Fixed `processData()` method
  EOF

  git commit -F /tmp/commit_msg

  Best practices for commit messages:
  - Use single quotes for simple messages with backticks
  - Use heredoc with single quotes for multi-line messages
  - Consider -F flag for complex messages to avoid shell escaping entirely

  The key is that commit messages often reference code with backticks, so single quotes or proper escaping prevents unintended command execution.

mattKorwel commented on packages/core/src/tools/shell.ts:
```diff
@@ -93,38 +81,198 @@ Process Group PGID: Process group started or \`(none)\``,
+   * @param command The shell command string to parse
+   * @returns An array of individual command strings
+   */
+  private splitCommands(command: string): string[] {
```
done. @scidomino I had to stomp on your windows changes during the merge commit. appologies. you'll need to revisit those changes, now in `shell-util.test.ts`
